### ISSUE TITLE
Simplify CI logs to improve debugging

### ISSUE THREAD

**Issue by praveenkuttappan** [Azure/azure-sdk-for-python#12536]
CI currently runs 5 different tests in parallel when running scheduled job and this makes it harder to find what exactly has failed due to current logging when tox is running in parallel. We need to simplify this logging to make it easier to find failures and reasons.

Please find below email thread for more info

```
Yes, that’s accurate. The invocation of tox is in serial (one package at a time), but the running of the environments for each package is in parallel. 

It’s dual core, so that means that 2 are invoking at a time.

If we want to split the steps up so that we have:

•	Run Dev Tests (sdist and whl)
•	Run Import Tests (depends)
•	Run MinMaxTests (mindep,maxdep)
•	Run Regression (currently another job, let’s leave this out FOR NOW)

So if we were to just change nothing else other than splitting up the invocations, we’ll end up hurting our invocation time. I don’t necessarily think it’ll be bad, but we definitely should try a prototype of this for say, storage. Here’s an example of just that. This isn’t ready for checkin, but will give us a good idea about how long it will extend the runtime.

I do think on the outset, we would be better served by doing the following. 

1.	Really clean up the test output. Tox itself shouldn’t echo any logs unless there is a failure. Attach tox logs to builds as artifact.
2.	Suppress error output until end, provide single formatted report at end of test phase

@Praveen Kuttappan any other thoughts on improvements to diagnosability?

Thanks,
Scott

From: Laurent Mazuel <lmazuel@microsoft.com> 
Sent: Tuesday, June 30, 2020 3:38 PM
To: Scott Beddall <scbedd@microsoft.com>; Praveen Kuttappan <Praveen.Marottickal@microsoft.com>
Cc: Scott Kurtzeborn <scotk@microsoft.com>
Subject: RE: CI kudos and retrospectives

Do I understand correctly that “tox -e whl,sdist,mindep,maxdep,imports” are ran in parallel right now? And that splitting in ADO steps would break that?


From: Scott Beddall <scbedd@microsoft.com> 
Sent: Tuesday, June 30, 2020 3:01 PM
To: Praveen Kuttappan <Praveen.Marottickal@microsoft.com>; Laurent Mazuel <lmazuel@microsoft.com>
Cc: Scott Kurtzeborn <scotk@microsoft.com>
Subject: RE: CI kudos and retrospectives

There are definitely improvements we can make as to the output. I can easily be sold on that 😊 

Currently, we test tox environments in parallel in the “Run-Tests” step, but invoke them in serial.

Essentially:

-	cd /sdk/storage/azure-storage-blobs 
-	tox -e whl,sdist,mindep,maxdep,imports

The only way we can split these up (which I do not protest against) is to add an `setup_execute_tests` for each one of these environments. That will have a fairly negative impact on our runtimes as implemented right now. We’ll need to figure out how to make multiple tox invocations run at the same time. 

Another thing I think we should do is to disable detailed logs (but attach them to the build output) unless there is an ERROR

The output of setup_execute_tests should also be more terse in general, outside of maybe “pip freeze” and the like.

Thanks,
Scott

From: Praveen Kuttappan <Praveen.Marottickal@microsoft.com> 
Sent: Tuesday, June 30, 2020 10:50 AM
To: Laurent Mazuel <lmazuel@microsoft.com>; Scott Beddall <scbedd@microsoft.com>
Cc: Scott Kurtzeborn <scotk@microsoft.com>
Subject: RE: CI kudos and retrospectives

Thanks for this great feedback. I will discuss with Scott and come up with a plan on your suggestions. To begin with, I will have a wiki created about all checks we do in Engsys for python.

Thanks and regards,
Praveen

From: Laurent Mazuel <lmazuel@microsoft.com> 
Sent: Tuesday, June 30, 2020 10:38 AM
To: Scott Beddall <scbedd@microsoft.com>; Praveen Kuttappan <Praveen.Marottickal@microsoft.com>
Cc: Scott Kurtzeborn <scotk@microsoft.com>
Subject: CI kudos and retrospectives

Hey Scott, Praveen,

Last night the regression “mindependency” CI found a regression that was merged in “azure-mgmt-core” part of a PR of yesterday. This was a 100% correct discovery, and I’m really relieved that the CI was there, especially since the impact would have been wide and include the China team. I thought it deserves an email to thank you again for that system, and re-iterate how I think this kind of CI steps are important for Python 😊

When you were investigated what happened with Izzy, I noted two things I wanted to discuss:
-	Izzy was not aware a regression system was in place, which makes the understanding of what happened confusing to her
-	Even when I told her, it’s lost in the middle of a step log, and she couldn’t find it.

I have a few suggestions I want to debate with you:
-	Could we make an inventory of what the engsys is doing:
o	What kind of checks exists (mypy, pylint, unittests, regression, shared req, etc.)
o	What kind of CI exists (PR, nightly scheduled, etc.)
o	Maybe a giant table to cross all that?
Maybe it exists already, but I wasn’t able to find it myself on the internal wiki if that’s the case.
-	Could we split CI in more steps. I understand the limitation of agents, but could we make more steps without making it slower? “Run Tests” to do both unittests + mindependency was confusing
-	Could we imagine a brown bag series once we get clarity of docs, to be sure the whole team is on the same page. I’m no fool, I’m fairly sure this would need to be done every 6 months or so to keep up and refresh people memory.

Again, thanks for the regression catch 😉
Let me know your thoughts,

Thanks!

Laurent
